### PR TITLE
Nursery: don't act as a checkpoint when not running anything

### PR DESCRIPTION
**Pull request by catern**
Previously, a Nursery would checkpoint on exit, regardless of whether
it was running anything. This is a bit annoying, see #1457, so let's
change it to not checkpoint on exit. Now it won't be a checkpoint at
all if there's no tasks running inside.

However, this doesn't fully solve #1457, because we'll still inject a
cancel even if none of the child tasks raise Cancelled.

This seems to break
trio/_core/tests/test_asyncgen.py::test_last_minute_gc_edge_case
such that it no longer ever hits the edge case.

That seems like a pretty complicated test, so maybe @oremanj can say
better why exactly it's breaking with this change.

### CODE REVIEW

**Comment by njsmith** [comment]
> However, this doesn't fully solve #1457, because we'll still inject a cancel even if none of the child tasks raise Cancelled.

Here's the code that does that, at line 923:

```python
            def aborted(raise_cancel):
                self._add_exc(capture(raise_cancel).error)
```

...I was going to say we could delete this line and simply ignore cancellation here, but that's not *quite* right. We do want to ignore `Cancelled` exceptions here, but if Trio is using cancellation to deliver a `KeyboardInterrupt` to the main task while it's blocked in `__aexit__`, then we still need that to be injected so that the nursery will cancel the other tasks and then raise `KeyboardInterrupt`. That should be fixed at some point Soon™, but in the mean time I guess we could work around this by adding a check for whether `capture(raise_cancel).error` is a `Cancelled`, and if so then ignore it.

**Comment by oremanj** [comment]
I left a comment on #1457: I think we shouldn't drop the schedule point in either branch, only the cancellation point; and we shouldn't drop the cancellation point if the nursery didn't run any tasks, so that `async with open_nursery(): pass` still executes a full checkpoint.

The test is failing because it relies on nursery exit being a schedule point.

File: trio/_core/tests/test_run.py [hunk 1]
```diff
@@ -1685,13 +1681,24 @@ async def nothing(task_status=_core.TASK_STATUS_IGNORED):
...
+    # but if the task does not execute any checkpoints, and exits, then start()
+    # doesn't raise Cancelled, since the task completed successfully.
+    async def started_with_no_checkpoint(task_status=_core.TASK_STATUS_IGNORED):
+        task_status.started("hi")
+
+    async with _core.open_nursery() as nursery:
+        with _core.CancelScope() as cs:
+            cs.cancel()
+            await nursery.start(started_with_no_checkpoint)
+
```
Comment (catern) [inline trio/_core/tests/test_run.py:1701]: Some more implications here. This is because a Nursery is used in the implementation of start().

I'd argue that this is correct.

**Comment by catern** [comment]
Also, what do you think about making assert_yields_or_not public? I just needed it for the test that a nursery is a schedule point but not a cancel point. Maybe that's something we should discourage?

**Comment by njsmith** [comment]
> Also, what do you think about making assert_yields_or_not public? I just needed it for the test that a nursery is a schedule point but not a cancel point. Maybe that's something we should discourage?

I actually just ran into this issue again while working on #1805 on stream, so I did some tweaks to land this quicker. And one of those tweaks was to revert the `assert_yields_or_not` changes, because they seemed overly complex for a single weird internal test :-). So I guess that answers the question...

File: newsfragments/1457.feature.rst [hunk 2]
```diff
@@ -0,0 +1,7 @@
+When exiting a nursery block, the parent task always waits for child
+tasks to exit. This wait cannot be cancelled. However, previously, if
+you tried to cancel it, it *would* inject a `Cancelled` exception,
+even though it wasn't cancelled. Most users probably never noticed
```
Comment (goodboy) [inline newsfragments/1457.feature.rst:4]: > This wait cannot be cancelled. However, previously, if
you tried to cancel it, it would inject a Cancelled exception,
even though it wasn't cancelled.

So this means the `Cancelled` is still injected but now is ignored via the new [`isinstance` check](https://github.com/python-trio/trio/pull/1696/files#diff-568b66d9830d9516b524413eb041a1d4c8ccf0b76af4e10d96098888ce323235R924)?

Also if the following is true:

> This wait cannot be cancelled. However, previously, if
you tried to cancel it, it would inject a Cancelled exception,
even though it wasn't cancelled.

Then why [the change to ` cancel_shielded_checkpoint()`](https://github.com/python-trio/trio/pull/1696/files#diff-568b66d9830d9516b524413eb041a1d4c8ccf0b76af4e10d96098888ce323235R935)?

File: trio/_core/tests/test_run.py [hunk 3]
```diff
@@ -433,12 +433,12 @@ async def crasher():
...
                 # nursery block continue propagating to reach the
                 # outer scope.
-                assert len(multi_exc.exceptions) == 5
+                assert len(multi_exc.exceptions) == 4
                 summary = {}
                 for exc in multi_exc.exceptions:
                     summary.setdefault(type(exc), 0)
                     summary[type(exc)] += 1
-                assert summary == {_core.Cancelled: 4, KeyError: 1}
+                assert summary == {_core.Cancelled: 3, KeyError: 1}
```
Comment (goodboy) [inline trio/_core/tests/test_run.py:441]: Clarifying the cancelled count in a comment might be clearer here for those unfamiliar with the recent change?

Something like *we only receive cancelleds from subtasks not from the nursery itself*?

File: trio/_core/tests/test_run.py [hunk 4]
```diff
@@ -1605,20 +1608,35 @@ async def test_trivial_yields():
...
+
+    before_schedule_points = task._schedule_points
+
+    async def noop_with_no_checkpoint():
+        pass
+
+    with _core.CancelScope() as cs:
+        cs.cancel()
+        async with _core.open_nursery() as nursery:
+            nursery.start_soon(noop_with_no_checkpoint)
```
Comment (goodboy) [inline trio/_core/tests/test_run.py:1630]: I presume it's out of scope (and already covered in other tests) but normally doing this cancellation before opening the nursery results in cancellation at the first checkpoint right?

File: trio/_core/tests/test_run.py [hunk 5]
```diff
@@ -1685,28 +1703,37 @@ async def nothing(task_status=_core.TASK_STATUS_IGNORED):
...
-    # and if after the no-op started(), the child crashes, the error comes out
-    # of start()
+    # but if the task does not execute any checkpoints, and exits, then start()
+    # doesn't raise Cancelled, since the task completed successfully.
+    async def started_with_no_checkpoint(task_status=_core.TASK_STATUS_IGNORED):
+        task_status.started("hi")
+
+    async with _core.open_nursery() as nursery:
+        with _core.CancelScope() as cs:
+            cs.cancel()
```
Comment (goodboy) [inline trio/_core/tests/test_run.py:1721]: In the case where you cancel from inside the nursery after having started tasks with checkpoints, won't counts of expected cancels change as well or is that entirely covered by the one test change in `test_run.py`?

**Review by goodboy — COMMENTED** [review]
Code all looks great (as usual) but some minor clarifications to an onlooker (newb) such as myself might be handy 😸

**Comment by catern** [comment]
The changes look fine to me. Excited for this to be fixed!

**Comment by richardsheridan** [comment]
The merge was not as painful as one might expect. The two new commits are nearly trivial fixes responsive to recent updates to the test suite.
